streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2472,18 +2472,58 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
     }
 
   /**
-   * Interrupts the stream if it does not produce a value after d duration.
+   * Ends the stream if it does not produce a value after d duration.
    */
   final def timeout(d: Duration): ZStream[R with Clock, E, O] =
     ZStream[R with Clock, E, O] {
       self.process.map { next =>
         next.timeout(d).flatMap {
-          case Some(a) => ZIO.succeedNow(a)
-          case None    => ZIO.interrupt
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.end
```

Comment:
We should add a Ref here that remembers that this stream has ended due to timeout.

Reply:
@iravid not sure what should be the behavior here, something like:

```scala
ZStream[R with Clock, E, O] {
  for {
    timeout <- Ref.make(false).toManaged_
    next    <- self.process
    pull = timeout.get.flatMap {
      if (_) Pull.end
      else
        next.timeout(d).flatMap {
          case Some(a) => Pull.emit(a)
          case None    => timeout.set(true) *> Pull.end
        }
    }
  } yield pull
}
```
?

Reply:
Yes, just like that.

Reply:
:+1: